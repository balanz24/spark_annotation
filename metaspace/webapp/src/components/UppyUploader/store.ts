// based on https://uppy.io/docs/stores/#Implementing-Stores

import { ref } from '@vue/composition-api'
import { State } from '@uppy/core'

type Listener = (prevState: State, nextState: State, patch: object) => void

export default () => {
  const state = ref<State>({})
  const listeners = new Set<Listener>()

  return {
    getState: () => state.value,
    setState: (patch: object) => {
      const prevState = state.value
      const nextState = { ...prevState, ...patch }

      state.value = nextState

      listeners.forEach(listener => {
        listener(prevState, nextState, patch)
      })
    },
    subscribe: (listener: Listener) => {
      listeners.add(listener)
      return () => listeners.delete(listener)
    },
  }
}
