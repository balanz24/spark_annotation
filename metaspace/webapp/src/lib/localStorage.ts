import safeJsonParse from './safeJsonParse'
import * as cookie from 'js-cookie'

const memoryStorage: Record<string, string> = {}

export const setLocalStorage = (key: string, value: any, cookieFallback = false) => {
  const json = JSON.stringify(value)
  memoryStorage[key] = json
  try {
    localStorage.setItem(key, json)
  } catch (err) {
    console.error(err)
    if (cookieFallback) {
      try {
        cookie.set('storage_' + key, json, { expires: 90 })
      } catch (err2) { console.error(err2) }
    }
  }
}

export const getLocalStorage = <T>(key: string): (T | undefined) => {
  try {
    const json = localStorage.getItem(key) || memoryStorage[key] || cookie.get('storage_' + key)
    return json && safeJsonParse(json)
  } catch (err) {
    console.error(err)
    return undefined
  }
}

export const removeLocalStorage = (key: string) => {
  delete memoryStorage[key]
  try {
    localStorage.removeItem(key)
  } catch (err) { console.error(err) }
  try {
    cookie.remove('storage_' + key)
  } catch (err) { console.error(err) }
}

export const migrateLocalStorage = () => {
  try {
    // Clean up old data stored by Upload page
    removeLocalStorage('latestMetadataSubmission')
    removeLocalStorage('latestMetadataOptions')
    removeLocalStorage('latestMetadataSubmissionVersion')

    // Convert storage cookies to localStorage
    try {
      const hideReleaseNotes = cookie.getJSON('hideReleaseNotes')
      if (hideReleaseNotes != null) {
        setLocalStorage('hideReleaseNotes', hideReleaseNotes, true)
        cookie.remove('hideReleaseNotes')
      }
    } catch (err) { console.error(err) }
    try {
      const showDescrHint = cookie.getJSON('show_descr_hint') as any
      if (showDescrHint != null) {
        setLocalStorage('hideMarkdownHint', showDescrHint === 0, true)
        cookie.remove('show_descr_hint')
      }
    } catch (err) { console.error(err) }
  } catch (err) { console.error(err) }
}
