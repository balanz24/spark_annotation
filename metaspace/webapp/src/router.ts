/* eslint-disable vue/max-len */
import Vue, { AsyncComponent } from 'vue'
import VueRouter, { RawLocation } from 'vue-router'
import AboutPage from './modules/App/AboutPage'
import DatasetsPage from './modules/Datasets/DatasetsPage.vue'
import { DialogPage, ResetPasswordPage } from './modules/Account'
import { redirectAfterSignIn } from './modules/Account/signInReturnUrl'
import { updateDBParam } from './modules/Filters/url'
import NotFoundPage from './modules/App/NotFoundPage.vue'

Vue.use(VueRouter)

const asyncPagesFreelyTyped = {
  AnnotationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "AnnotationsPage" */ './modules/Annotations/AnnotationsPage.vue'),
  DatasetSummary: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetSummary" */ './modules/Datasets/summary/DatasetSummary.vue'),
  MetadataEditPage: () => import(/* webpackPrefetch: true, webpackChunkName: "MetadataEditPage" */ './modules/MetadataEditor/MetadataEditPage.vue'),
  ImageAlignmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "ImageAlignmentPage" */ './modules/ImageAlignment/ImageAlignmentPage.vue'),
  UploadPage: () => import(/* webpackPrefetch: true, webpackChunkName: "UploadPage" */ './modules/MetadataEditor/UploadPage.vue'),
  DatasetOverviewPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/overview/DatasetOverviewPage'),
  DatasetComparisonPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/comparison/DatasetComparisonPage'),
  DatasetBrowserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/imzml/DatasetBrowserPage'),
  DatasetEnrichmentPage: () => import(/* webpackPrefetch: true, webpackChunkName: "DatasetOverviewPage" */ './modules/Datasets/enrichment/DatasetEnrichmentPage'),
  SpottingProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "SpottingProjectPage" */ './modules/SpottingProject/DashboardPage'),

  // These pages are relatively small as they don't have any big 3rd party dependencies, so pack them together
  DatasetTable: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Datasets/list/DatasetTable.vue'),
  HelpPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/HelpPage.vue'),
  EditUserPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/UserProfile/EditUserPage.vue'),
  CreateGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/CreateGroupPage.vue'),
  ProjectsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Project/ProjectsListPage.vue'),
  SystemHealthPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Admin/SystemHealthPage.vue'),
  GroupsListPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/Group/GroupsListPage'),
  PrivacyPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PrivacyPage.vue'),
  TermsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/TermsPage.vue'),
  PublicationsPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle1" */ './modules/App/PublicationsPage'),

  // These pages use sanitizeHtml, which is big
  ViewGroupPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Group/ViewGroupPage.vue'),
  ViewProjectPage: () => import(/* webpackPrefetch: true, webpackChunkName: "Bundle2" */ './modules/Project/ViewProjectPage.vue'),

  // Separate bundle for design docs
  DesignTOC: () => import(/* webpackChunkName: "DesignBundle" */ './design/TOCPage.vue'),
  DesignStyleGuide: () => import(/* webpackChunkName: "DesignBundle" */ './design/StyleGuidePage.vue'),
  DesignIcons: () => import(/* webpackChunkName: "DesignBundle" */ './design/IconsPage.vue'),
  DesignComponents: () => import(/* webpackChunkName: "DesignBundle" */ './design/ComponentsPage.vue'),
  DesignForms: () => import(/* webpackChunkName: "DesignBundle" */ './design/FormsPage.vue'),
}
const asyncPages = asyncPagesFreelyTyped as Record<keyof typeof asyncPagesFreelyTyped, AsyncComponent>

const convertLegacyUrls = () => {
  const { pathname, hash, search } = window.location
  if (pathname === '/' && hash && hash.startsWith('#/')) {
    history.replaceState(undefined, undefined as any, hash.slice(1))
  }
  if (pathname === '/annotations') {
    const updatedQueryString = updateDBParam(search.slice(1))
    if (updatedQueryString !== null) {
      history.replaceState(undefined, undefined as any, `${pathname}?${updatedQueryString}`)
    }
  }
}
convertLegacyUrls()

const router = new VueRouter({
  mode: 'history',
  routes: [
    { path: '/(about)?', component: AboutPage, meta: { footer: true, headerClass: 'bg-primary' } },
    { path: '/annotations', name: 'annotations', component: asyncPages.AnnotationsPage },
    {
      path: '/datasets',
      component: DatasetsPage,
      children: [
        { path: '', component: asyncPages.DatasetTable },
        { path: 'summary', component: asyncPages.DatasetSummary },
      ],
    },
    {
      path: '/datasets/:dataset_id/comparison',
      name: 'datasets-comparison',
      component: asyncPages.DatasetComparisonPage,
    },
    { path: '/datasets/edit/:dataset_id', name: 'edit-metadata', component: asyncPages.MetadataEditPage },
    { path: '/datasets/:dataset_id/add-optical-image', name: 'add-optical-image', component: asyncPages.ImageAlignmentPage },
    { path: '/dataset/:dataset_id', name: 'dataset-overview', component: asyncPages.DatasetOverviewPage },
    { path: '/dataset/:dataset_id/annotations', name: 'dataset-annotations', component: asyncPages.AnnotationsPage },
    { path: '/dataset/:dataset_id/browser', name: 'dataset-browser', component: asyncPages.DatasetBrowserPage },
    { path: '/dataset/:dataset_id/enrichment', name: 'dataset-enrichment', component: asyncPages.DatasetEnrichmentPage },
    { path: '/upload', component: asyncPages.UploadPage },
    { path: '/help', component: asyncPages.HelpPage, meta: { footer: true } },
    { path: '/user/me', component: asyncPages.EditUserPage },

    { path: '/admin/health', component: asyncPages.SystemHealthPage },

    { path: '/account/sign-in', component: DialogPage, props: { dialog: 'signIn' } },
    { path: '/account/sign-in-success', redirect: redirectAfterSignIn },
    { path: '/account/create-account', component: DialogPage, props: { dialog: 'createAccount' } },
    { path: '/account/forgot-password', component: DialogPage, props: { dialog: 'forgotPassword' } },
    { path: '/account/reset-password', component: ResetPasswordPage },

    { path: '/groups', component: asyncPages.GroupsListPage },
    { path: '/group/create', component: asyncPages.CreateGroupPage },
    { path: '/group/:groupIdOrSlug', name: 'group', component: asyncPages.ViewGroupPage },

    { path: '/project/:projectIdOrSlug', name: 'project', component: asyncPages.ViewProjectPage },
    { // Legacy URL sent in "request access" emails up until Feb 2019
      path: '/project/:projectIdOrSlug/manage',
      redirect: { path: '/project/:projectIdOrSlug', query: { tab: 'members' } } as RawLocation,
    },
    { path: '/projects', component: asyncPages.ProjectsListPage },

    { path: '/terms', component: asyncPages.TermsPage, meta: { footer: true } },
    { path: '/privacy', component: asyncPages.PrivacyPage, meta: { footer: true } },
    { path: '/publications', component: asyncPages.PublicationsPage, meta: { footer: true } },

    { path: '/design', component: asyncPages.DesignTOC, meta: { footer: true, flex: true } },
    { path: '/design/styleguide', component: asyncPages.DesignStyleGuide, meta: { footer: true, flex: true } },
    { path: '/design/icons', component: asyncPages.DesignIcons, meta: { footer: true, flex: true } },
    { path: '/design/components', component: asyncPages.DesignComponents, meta: { footer: true, flex: true } },
    { path: '/design/forms', component: asyncPages.DesignForms, meta: { footer: true, flex: true } },

    { path: '/detectability', name: 'detectability', component: asyncPages.SpottingProjectPage },

    { path: '*', component: NotFoundPage, meta: { footer: true, flex: true } },
  ],
  scrollBehavior(to, from, savedPosition) {
    if (to.hash) {
      return { selector: to.hash, offset: { x: 0, y: 64 } } // offset header height
    }
    if (savedPosition) {
      return savedPosition
    }
    if (to.path !== from.path) {
      return { x: 0, y: 0 }
    }
  },
})

const pageLoadedAt = Date.now()
router.beforeEach((to, from, next) => {
  // If user has had the window open for a long time, it's likely they're using old code. This can cause issues
  // such as incorrect API usage, or causing error reports for bugs that have already been fixed.
  // Force the browser to reload the page, preferring to do so during page transitions to minimize impact.
  const daysSincePageLoad = (Date.now() - pageLoadedAt) / 1000 / 60 / 60 / 24
  if (to.path !== from.path && daysSincePageLoad > 1.75 || daysSincePageLoad > 7) {
    window.location.assign(to.fullPath)
    next(false)
  } else {
    next()
  }
})

const { href } = router.resolve({ name: 'project', params: { projectIdOrSlug: 'REMOVE' } }, undefined, true)
export const PROJECT_URL_PREFIX = location.origin + href.replace('REMOVE', '')

export default router
