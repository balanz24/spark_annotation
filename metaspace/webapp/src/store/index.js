import Vue from 'vue';
import Vuex from 'vuex';
Vue.use(Vuex);

import getters from './getters.js';
import mutations from './mutations.js';
import actions from './actions.js';
import {accountModule} from '../modules/Account';

const store = new Vuex.Store({
  state: {
    // names of currently shown filters
    orderedActiveFilters: [],

    filterLists: null,
    filterListsLoading: false,

    // currently selected annotation
    annotation: undefined,

    // currently selected normalization matrix
    normalization: undefined,

    // roi settings
    roiInfo: {visible: false},

    // is annotation table loading?
    tableIsLoading: true,

    lastUsedFilters: {},

    currentTour: null,

    currentUser: {},

    // ion image global viewer settings
    channels: [],
    mode: 'SINGLE'
  },

  getters,
  mutations,
  actions,
  modules: {
    account: accountModule
  }
})

export default store;
