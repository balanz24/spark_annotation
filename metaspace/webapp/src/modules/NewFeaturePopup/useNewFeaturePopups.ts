import { reactive, computed, ref } from '@vue/composition-api'
import { isArray } from 'lodash-es'

import { getLocalStorage, setLocalStorage } from '../../lib/localStorage'

const STORAGE_KEY = 'dismissedFeaturePopups'

export const FEATURE_KEYS = [
  'uploadCustomDatabases',
  'groupDatabasesTab',
  'multipleIonImages',
  'v2',
]

function getDismissedPopups() {
  try {
    const list = getLocalStorage(STORAGE_KEY)
    if (isArray(list)) {
      // Filter out invalid/old entries
      return list.filter(item => FEATURE_KEYS.includes(item))
    } else {
      return []
    }
  } catch (ex) {
    return []
  }
}

const previousDismissedPopups = getDismissedPopups()

const state = reactive({
  dismissed: [...previousDismissedPopups],
  queued: [] as (string | null)[],
})

const activePopup = computed(() => state.queued[0])

function closeActivePopup() {
  const popup = state.queued.shift()
  state.dismissed.push(popup)
  // WORKAROUND: Popper doesn't handle multiple instances - trying to open one Popper at the same time as closing
  // the last one will cause layout to break. Use a placeholder item for a few frames to allow the old Popper to close
  // before a new one is opened.
  state.queued.unshift(null)
  setTimeout(() => {
    if (state.queued[0] === null) { state.queued.shift() }
  }, 50)
  return popup
}

const popoverRef = ref<HTMLElement>()

export default () => {
  return {
    activePopup,
    popoverRef,
    isDismissed(featureKey: string) {
      return state.dismissed.includes(featureKey)
    },
    queuePopup(featureKey: string) {
      if (state.dismissed.includes(featureKey) || state.queued.includes(featureKey)) {
        return
      }
      state.queued.push(featureKey)
    },
    unqueuePopup(featureKey: string) {
      const index = state.queued.indexOf(featureKey)
      if (index !== -1) {
        state.queued.splice(index, 1)
      }
    },
    remindLater() {
      closeActivePopup()
    },
    dismissPopup() {
      const popup = closeActivePopup()
      previousDismissedPopups.push(popup)
      setLocalStorage(STORAGE_KEY, previousDismissedPopups)
    },
  }
}
