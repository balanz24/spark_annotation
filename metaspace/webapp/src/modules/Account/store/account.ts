import { Module } from 'vuex'
import { DialogType } from '../dialogs'
import router from '../../../router'
import { RawLocation } from 'vue-router/types/router'

export interface AccountState {
  dialog: DialogType | null;
  dialogCloseRedirect: RawLocation | null;
  loginSuccessRedirect: RawLocation | null;
}

interface ShowDialogOptions {
  dialog: DialogType;
  dialogCloseRedirect?: RawLocation;
  loginSuccessRedirect?: RawLocation;
}

interface HideDialogOptions {
  dialog: DialogType;
  isLoginSuccess?: boolean;
}

const account: Module<AccountState, any> = {
  namespaced: true,
  state: {
    dialog: null,
    dialogCloseRedirect: null,
    loginSuccessRedirect: null,
  },
  mutations: {
    showDialog(state, payload: DialogType | ShowDialogOptions) {
      const options: ShowDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      state.dialog = options.dialog
      if (options.dialogCloseRedirect != null) {
        state.dialogCloseRedirect = options.dialogCloseRedirect
      }
      if (options.loginSuccessRedirect != null) {
        state.loginSuccessRedirect = options.loginSuccessRedirect
      }
    },
    hideDialog(state, payload: DialogType | HideDialogOptions) {
      const options: HideDialogOptions = typeof payload === 'string' ? { dialog: payload } : payload

      if (state.dialog === options.dialog) {
        state.dialog = null
        if (options.isLoginSuccess && state.loginSuccessRedirect != null) {
          router.push(state.loginSuccessRedirect)
        } else if (state.dialogCloseRedirect != null) {
          router.push(state.dialogCloseRedirect)
        }

        state.dialogCloseRedirect = null
        state.loginSuccessRedirect = null
      }
    },
  },
}

export default account
