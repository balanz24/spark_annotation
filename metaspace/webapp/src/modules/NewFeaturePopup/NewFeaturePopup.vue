<template>
  <div
    v-if="isActive"
    ref="popover"
    class="el-popover el-popper leading-5 p-5 text-left"
  >
    <h3 class="leading-10 m-0 mt-2">
      <el-badge value="New">
        {{ title }}
      </el-badge>
    </h3>
    <div class="sm-content font-normal">
      <slot name="default" />
    </div>
    <div class="flex justify-end items-center h-10 mt-5">
      <el-button
        size="small"
        @click.stop="remindLater"
      >
        Remind me later
      </el-button>
      <el-button
        size="small"
        type="primary"
        @click.stop="dismissPopup"
      >
        Got it!
      </el-button>
    </div>
    <div
      class="popper__arrow"
      x-arrow=""
    />
  </div>
</template>
<script lang="ts">
import { defineComponent, computed, ref, onMounted, onBeforeUnmount, watch, onUnmounted } from '@vue/composition-api'

import useNewFeaturePopups from './useNewFeaturePopups'

interface Props {
  featureKey: string
  title: string
}

export default defineComponent<Props>({
  props: {
    featureKey: { type: String },
    title: String,
  },
  setup(props, { root }) {
    const {
      activePopup,
      popoverRef,
      dismissPopup,
      remindLater,
    } = useNewFeaturePopups()

    const isActive = computed(() => activePopup.value === props.featureKey)

    return {
      popover: popoverRef,
      isActive,
      remindLater,
      dismissPopup,
    }
  },
})
</script>
<style scoped>
  .el-popover {
    width: theme('maxWidth.sm')
  }

  /deep/ .sm-content > * {
    margin: 0;
  }
  /deep/ .sm-content > * + * {
    @apply mt-5;
  }
</style>
