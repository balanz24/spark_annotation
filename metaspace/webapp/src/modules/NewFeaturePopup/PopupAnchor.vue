<template>
  <span
    ref="reference"
    v-bind="$attrs"
  >
    <slot />
  </span>
</template>
<script lang="ts">
import { defineComponent, computed, ref, onMounted, onBeforeUnmount, watch, onUnmounted } from '@vue/composition-api'
import Popper, { Placement } from 'popper.js'

import useNewFeaturePopups from './useNewFeaturePopups'
import config from '../../lib/config'
import useIntersectionObserver from '../../lib/useIntersectionObserver'

interface Props {
  featureKey: string
  placement: Placement
  showUntil: Date
}

export default defineComponent<Props>({
  props: {
    featureKey: String,
    placement: String,
    showUntil: Date,
  },
  setup(props, { root }) {
    const {
      activePopup,
      isDismissed,
      queuePopup,
      unqueuePopup,
      popoverRef,
    } = useNewFeaturePopups()

    const reference = ref<HTMLElement>()

    const isRelevant = (props.showUntil || Number.MAX_VALUE) > new Date()

    if (!isRelevant || isDismissed(props.featureKey)) {
      return {}
    }

    const { isFullyInView, intersectionRatio } = useIntersectionObserver(reference, { threshold: [0, 1] })

    watch(isFullyInView, value => {
      if (value) {
        queuePopup(props.featureKey)
      } else if (activePopup.value !== props.featureKey) { // do not unqueue if popup is visible
        unqueuePopup(props.featureKey)
      }
    })

    const isActive = computed(() => {
      return (
        reference.value
        && popoverRef.value
        && config.features.new_feature_popups
        && root.$store.state.currentTour === null
        && activePopup.value === props.featureKey
      )
    })

    let popper : Popper | null

    watch(isActive, value => {
      if (value === true) {
        popper = new Popper(reference.value!, popoverRef.value!, { placement: props.placement })
      } else if (popper) {
        popper.destroy()
        popper = null
      }
    })

    onBeforeUnmount(() => {
      if (popper) {
        popper.destroy()
      }
      // in case popup was not dismissed
      unqueuePopup(props.featureKey)
    })

    return {
      reference,
    }
  },
})
</script>
