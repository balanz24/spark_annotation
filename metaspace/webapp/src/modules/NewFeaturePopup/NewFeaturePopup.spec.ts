import { mount } from '@vue/test-utils'
import Vue from 'vue'
import Vuex from 'vuex'
import { sync } from 'vuex-router-sync'

import NewFeaturePopup from './NewFeaturePopup.vue'
import PopupAnchor from './PopupAnchor.vue'
import useNewFeaturePopups from './useNewFeaturePopups'

import router from '../../router'
import store from '../../store'

// for mocks
import { ref } from '@vue/composition-api'
import * as useIntersectionObserver from '../../lib/useIntersectionObserver'
import * as localStorage from '../../lib/localStorage'

jest.mock('popper.js')
jest.mock('../../lib/useIntersectionObserver')
const mockUseIntersectionObserver = useIntersectionObserver as jest.Mocked<typeof useIntersectionObserver>
jest.mock('../../lib/localStorage')
const mockLocalStorage = localStorage as jest.Mocked<typeof localStorage>

Vue.use(Vuex)
sync(store, router)

function mockIntersectionObserver({ isIntersecting = false, isFullyInView = false, intersectionRatio = 0 }) {
  mockUseIntersectionObserver.default.mockImplementation(() => ({
    isIntersecting: ref(isIntersecting),
    isFullyInView: ref(isFullyInView),
    intersectionRatio: ref(intersectionRatio),
    observe: jest.fn(),
    unobserve: jest.fn(),
  }))
}

describe('NewFeaturePopup', () => {
  const TestNewFeaturePopup = Vue.component('test-new-feature-popup', {
    components: {
      anchor: PopupAnchor, // full name is mocked
      NewFeaturePopup,
    },
    props: ['showUntil'],
    template: `
      <div>
        <anchor feature-key="test" :show-until="showUntil">
          <span>Test anchor</span>
        </anchor>
        <new-feature-popup ref="popup" title="Test title" feature-key="test">
          <p>test content</p>
        </new-feature-popup>
      </div>
    `,
  })

  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should match snapshot (not in view)', async() => {
    mockIntersectionObserver({
      isFullyInView: false,
    })
    const wrapper = mount(TestNewFeaturePopup, { store, router })
    await Vue.nextTick()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should match snapshot (in view)', async() => {
    mockIntersectionObserver({
      isFullyInView: true,
    })
    const wrapper = mount(TestNewFeaturePopup, { store, router })
    await Vue.nextTick()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should close on remind later', async() => {
    mockIntersectionObserver({
      isFullyInView: true,
    })
    const { remindLater } = useNewFeaturePopups()

    const wrapper = mount(TestNewFeaturePopup, { store, router })
    await Vue.nextTick()

    remindLater()
    await Vue.nextTick()

    const popup = wrapper.findComponent(NewFeaturePopup)
    expect(popup.html()).toBe('')
  })

  it('should not show if dismissed', async() => {
    mockIntersectionObserver({
      isFullyInView: true,
    })
    mockLocalStorage.getLocalStorage.mockImplementation(() => ['test'])

    const wrapper = mount(TestNewFeaturePopup, { store, router })
    await Vue.nextTick()

    const popup = wrapper.findComponent(NewFeaturePopup)
    expect(popup.html()).toBe('')
  })
})
