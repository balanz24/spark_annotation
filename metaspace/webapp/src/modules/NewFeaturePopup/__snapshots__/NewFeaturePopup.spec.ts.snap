// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewFeaturePopup should match snapshot (in view) 1`] = `
<div>
  <span>
    <span>
      Test anchor
    </span>
  </span>
   
  <div
    class="el-popover el-popper leading-5 p-5 text-left"
  >
    <h3
      class="leading-10 m-0 mt-2"
    >
      <div
        class="el-badge"
      >
        
      Test title
    
        <transition-stub
          name="el-zoom-in-center"
        >
          <sup
            class="el-badge__content el-badge__content--undefined is-fixed"
          >
            New
          </sup>
        </transition-stub>
      </div>
    </h3>
     
    <div
      class="sm-content font-normal"
    >
      <p>
        test content
      </p>
    </div>
     
    <div
      class="flex justify-end items-center h-10 mt-5"
    >
      <button
        class="el-button el-button--default el-button--small"
        type="button"
      >
        <!---->
        <!---->
        <span>
          
      Remind me later
    
        </span>
      </button>
       
      <button
        class="el-button el-button--primary el-button--small"
        type="button"
      >
        <!---->
        <!---->
        <span>
          
      Got it!
    
        </span>
      </button>
    </div>
     
    <div
      class="popper__arrow"
      x-arrow=""
    />
  </div>
</div>
`;

exports[`NewFeaturePopup should match snapshot (not in view) 1`] = `
<div>
  <span>
    <span>
      Test anchor
    </span>
  </span>
   
  <!---->
</div>
`;
